File: frontend/src/lib/api/tickets.ts
/**
 * Tickets API service
 */

import { apiClient } from "./client";
import { transformTicket, transformTickets, transformTicketStats, toSnakeCase } from "./transformers";
import type { Ticket, TicketStats, TicketStatus, TicketType } from "@/types";
import type { TicketApiResponse, TicketStatsApiResponse, TicketPayRequest } from "./types";

export interface TicketListParams {
  skip?: number;
  limit?: number;
  status?: TicketStatus;
  type?: TicketType;
  vehicle_id?: string;
  driver_id?: string;
  overdue_only?: boolean;
  start_date?: string;
  end_date?: string;
}

export const ticketsService = {
  /**
   * Get all tickets with optional filters
   */
  async getAll(params?: TicketListParams): Promise<Ticket[]> {
    const queryParams: Record<string, string | number | boolean | undefined> = {
      skip: params?.skip,
      limit: params?.limit,
      status: params?.status,
      type: params?.type,
      vehicle_id: params?.vehicle_id,
      employee_id: params?.driver_id,
      overdue_only: params?.overdue_only,
      start_date: params?.start_date,
      end_date: params?.end_date,
    };

    const response = await apiClient.get<TicketApiResponse[]>("/tickets", {
      params: queryParams,
    });

    return transformTickets(response);
  },

  /**
   * Get ticket by ID
   */
  async getById(id: string): Promise<Ticket> {
    const response = await apiClient.get<TicketApiResponse>(`/tickets/${id}`);
    return transformTicket(response);
  },

  /**
   * Create a new ticket
   */
  async create(data: Partial<Ticket>): Promise<Ticket> {
    const request = toSnakeCase(data);
    const response = await apiClient.post<TicketApiResponse>("/tickets", request);
    return transformTicket(response);
  },

  /**
   * Update a ticket
   */
  async update(id: string, data: Partial<Ticket>): Promise<Ticket> {
    const request = toSnakeCase(data);
    const response = await apiClient.patch<TicketApiResponse>(`/tickets/${id}`, request);
    return transformTicket(response);
  },

  /**
   * Delete a ticket
   */
  async delete(id: string): Promise<void> {
    await apiClient.delete(`/tickets/${id}`);
  },

  /**
   * Mark ticket as paid
   */
  async markAsPaid(id: string, paymentData: {
    paidDate: string;
    paidAmount: number;
    paymentMethod: string;
    paymentReference?: string;
  }): Promise<Ticket> {
    const request: TicketPayRequest = {
      paid_date: paymentData.paidDate,
      paid_amount: paymentData.paidAmount,
      payment_method: paymentData.paymentMethod,
      payment_reference: paymentData.paymentReference,
    };
    const response = await apiClient.post<TicketApiResponse>(`/tickets/${id}/pay`, request);
    return transformTicket(response);
  },

  /**
   * Update ticket status
   */
  async updateStatus(id: string, newStatus: TicketStatus): Promise<Ticket> {
    const response = await apiClient.post<TicketApiResponse>(
      `/tickets/${id}/status`,
      undefined,
      { params: { new_status: newStatus } }
    );
    return transformTicket(response);
  },

  /**
   * Get overdue tickets
   */
  async getOverdue(skip?: number, limit?: number): Promise<Ticket[]> {
    const response = await apiClient.get<TicketApiResponse[]>("/tickets/overdue", {
      params: { skip, limit },
    });
    return transformTickets(response);
  },

  /**
   * Get ticket statistics
   */
  async getStats(startDate?: string, endDate?: string): Promise<TicketStats> {
    const response = await apiClient.get<TicketStatsApiResponse>("/tickets/stats", {
      params: { start_date: startDate, end_date: endDate },
    });
    return transformTicketStats(response);
  },

  /**
   * Get tickets by vehicle
   */
  async getByVehicle(vehicleId: string, skip?: number, limit?: number): Promise<Ticket[]> {
    const response = await apiClient.get<TicketApiResponse[]>(`/tickets/by-vehicle/${vehicleId}`, {
      params: { skip, limit },
    });
    return transformTickets(response);
  },

  /**
   * Get tickets by driver
   */
  async getByDriver(driverId: string, skip?: number, limit?: number): Promise<Ticket[]> {
    const response = await apiClient.get<TicketApiResponse[]>(`/tickets/by-employee/${driverId}`, {
      params: { skip, limit },
    });
    return transformTickets(response);
  },
};

export default ticketsService;

File: backend/app/modules/fleet/api/tickets.py
"""Ticket endpoints."""

from datetime import date

from fastapi import APIRouter

from app.shared.api.deps import CurrentUserDep, DBDep, FleetManagerDep
from app.models.enums import TicketStatus, TicketType
from app.schemas import (
    TicketCreate,
    TicketPayRequest,
    TicketResponse,
    TicketStats,
    TicketUpdate,
)
from app.services import ticket_service

router = APIRouter()


@router.get("", response_model=list[TicketResponse])
def list_tickets(
    db: DBDep,
    current_user: CurrentUserDep,
    skip: int = 0,
    limit: int = 100,
    status: TicketStatus | None = None,
    type: TicketType | None = None,
    vehicle_id: str | None = None,
    employee_id: str | None = None,
    overdue_only: bool = False,
    start_date: date | None = None,
    end_date: date | None = None,
) -> list[TicketResponse]:
    """List tickets in the organization with optional filters."""
    tickets = ticket_service.get_multi_filtered(
        db,
        organization_id=current_user.organization_id,
        status=status,
        ticket_type=type,
        vehicle_id=vehicle_id,
        employee_id=employee_id,
        overdue_only=overdue_only,
        start_date=start_date,
        end_date=end_date,
        skip=skip,
        limit=limit,
    )
    return [TicketResponse.model_validate(t) for t in tickets]


@router.post("", response_model=TicketResponse, status_code=201)
def create_ticket(
    ticket_in: TicketCreate,
    db: DBDep,
    current_user: FleetManagerDep,
) -> TicketResponse:
    """Create a new ticket."""
    ticket_data = ticket_in.model_copy(update={"organization_id": current_user.organization_id})
    ticket = ticket_service.create(db, obj_in=ticket_data)
    return TicketResponse.model_validate(ticket)


@router.get("/overdue", response_model=list[TicketResponse])
def list_overdue_tickets(
    db: DBDep,
    current_user: CurrentUserDep,
    skip: int = 0,
    limit: int = 100,
) -> list[TicketResponse]:
    """List overdue tickets (past due date and not paid/cancelled)."""
    tickets = ticket_service.get_overdue(
        db,
        organization_id=current_user.organization_id,
        skip=skip,
        limit=limit,
    )
    return [TicketResponse.model_validate(t) for t in tickets]


@router.get("/stats", response_model=TicketStats)
def get_ticket_stats(
    db: DBDep,
    current_user: CurrentUserDep,
    start_date: date | None = None,
    end_date: date | None = None,
) -> TicketStats:
    """Get ticket statistics for the organization."""
    return ticket_service.get_stats(
        db,
        organization_id=current_user.organization_id,
        start_date=start_date,
        end_date=end_date,
    )


@router.get("/by-vehicle/{vehicle_id}", response_model=list[TicketResponse])
def list_tickets_by_vehicle(
    vehicle_id: str,
    db: DBDep,
    current_user: CurrentUserDep,
    skip: int = 0,
    limit: int = 100,
) -> list[TicketResponse]:
    """List all tickets for a specific vehicle."""
    tickets = ticket_service.get_by_vehicle(
        db,
        vehicle_id=vehicle_id,
        organization_id=current_user.organization_id,
        skip=skip,
        limit=limit,
    )
    return [TicketResponse.model_validate(t) for t in tickets]


@router.get("/by-employee/{employee_id}", response_model=list[TicketResponse])
def list_tickets_by_employee(
    employee_id: str,
    db: DBDep,
    current_user: CurrentUserDep,
    skip: int = 0,
    limit: int = 100,
) -> list[TicketResponse]:
    """List all tickets for a specific employee."""
    tickets = ticket_service.get_by_employee(
        db,
        employee_id=employee_id,
        organization_id=current_user.organization_id,
        skip=skip,
        limit=limit,
    )
    return [TicketResponse.model_validate(t) for t in tickets]


@router.get("/{ticket_id}", response_model=TicketResponse)
def get_ticket(
    ticket_id: str,
    db: DBDep,
    current_user: CurrentUserDep,
) -> TicketResponse:
    """Get a ticket by ID."""
    ticket = ticket_service.get_or_404(db, ticket_id)
    return TicketResponse.model_validate(ticket)


@router.patch("/{ticket_id}", response_model=TicketResponse)
def update_ticket(
    ticket_id: str,
    ticket_in: TicketUpdate,
    db: DBDep,
    current_user: FleetManagerDep,
) -> TicketResponse:
    """Update a ticket."""
    ticket = ticket_service.get_or_404(db, ticket_id)
    updated = ticket_service.update(db, db_obj=ticket, obj_in=ticket_in)
    return TicketResponse.model_validate(updated)


@router.delete("/{ticket_id}", status_code=204)
def delete_ticket(
    ticket_id: str,
    db: DBDep,
    current_user: FleetManagerDep,
) -> None:
    """Delete a ticket (soft delete)."""
    ticket_service.delete(db, ticket_id)


@router.post("/{ticket_id}/pay", response_model=TicketResponse)
def pay_ticket(
    ticket_id: str,
    payment_in: TicketPayRequest,
    db: DBDep,
    current_user: FleetManagerDep,
) -> TicketResponse:
    """Mark a ticket as paid with payment details."""
    ticket = ticket_service.get_or_404(db, ticket_id)
    updated = ticket_service.mark_as_paid(db, ticket, payment_in)
    return TicketResponse.model_validate(updated)


@router.post("/{ticket_id}/status", response_model=TicketResponse)
def update_ticket_status(
    ticket_id: str,
    new_status: TicketStatus,
    db: DBDep,
    current_user: FleetManagerDep,
) -> TicketResponse:
    """Update ticket status (e.g., to appealed or cancelled)."""
    ticket = ticket_service.get_or_404(db, ticket_id)
    updated = ticket_service.update_status(db, ticket, new_status)
    return TicketResponse.model_validate(updated)
